### ISSUE TITLE
GeoArray isn't a concrete type

### ISSUE THREAD

**Issue by evetion** [evetion/GeoArrays.jl#80]
A::AbstractArray{T,3}
    f::AffineMap
    crs::WellKnownText{GeoFormatTypes.CRS,<:String}
> Hmm none of the three fields are concrete types, right? Is there a reason you went with that? Since normally it's best to avoid such that the compiler can specialize: https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-fields-with-abstract-type

_Originally posted by @visr in https://github.com/evetion/GeoArrays.jl/pull/79#discussion_r702466874_

**Comment by evetion** [comment]
Proper types have never been my strong suit. @visr what do you think of the following?

```julia
mutable struct GeoArray{T<:Union{Real,Union{Missing,Real}},A<:AbstractArray{T,3},AM<:AbstractAffineMap,WKT<:GeoFormatTypes.AbstractWellKnownText} <: AbstractArray{T,3}
   A::A
   f::AM
   crs::WKT
end
```

```julia
julia> using GeoArrays

julia> ga = GeoArray(rand(5,5))
5x5x1 Array{Float64, 3} with AffineMap([1.0 0.0; 0.0 1.0], [0.0, 0.0]) and undefined CRS

julia> typeof(ga)
GeoArray{Float64, Array{Float64, 3}, CoordinateTransformations.AffineMap{StaticArrays.SMatrix{2, 2, Float64, 4}, StaticArrays.SVector{2, Float64}}, GeoFormatTypes.WellKnownText{GeoFormatTypes.CRS, String}}

julia> isconcretetype(typeof(ga))
true
```

**Comment by visr** [comment]
Yeah that looks pretty good. What is the reason for parameterizing the AffineMap, are the multiple ones that you want to use? If not, I'd just fix it. Same perhaps for WKT?

For the array when reading I guess it would normally be a DiskArray from ArchGDAL.readraster, and (often) an Array when users create it themselves.

**Comment by evetion** [comment]
I thought it was required, as the underlying types are parametrized as well, but it seems I was wrong. See updated PR.

**Comment by visr** [comment]
Yeah indeed the underlying types are parametrized. But there are two ways to handle that, (1) to add those parameters to your wrapping type, GeoArray, (2) to not support all `AffineMap` or `GeoFormatTypes.AbstractWellKnownText`, but just `AffineMap{SMatrix{2, 2, Float64, 4}, SVector{2, Float64}}` and `WellKnownText2{CRS, String}` for instance. If you want to support EPSG as well, then during construction you can use ArchGDAL's convert methods to convert to WKT2.

The `isconcretetype` check here should be applied to the fields, not GeoArray:

```
julia> isconcretetype(Vector{Any})
true

julia> isconcretetype(AffineMap)
false
```

**Comment by visr** [comment]
It's not yet fixed right? Though now 1 of the 3 fields is concrete, see my comment above.

**Comment by evetion** [comment]
Well, non initialized structs are never concrete (check `isconcretetype(Vector)`). As soon as you do:

```julia
julia> AffineMap([1 1;1 1],[1,1])
AffineMap([1 1; 1 1], [1, 1])

julia> typeof(AffineMap([1 1;1 1],[1,1]))
AffineMap{Matrix{Int64}, Vector{Int64}}

julia> isconcretetype(typeof(AffineMap([1 1;1 1],[1,1])))
true
```

I've added tests for the default constructor that confirm this:
https://github.com/evetion/GeoArrays.jl/pull/85/files#diff-a562daaa67f3cc5c1f500a2abbb46325428ef968d5e252f8122448ad2041d4f9R11

**Comment by visr** [comment]
Yeah but there you are checking an instantiated type, which by definition is concrete, in the same sense that there is no number that is Real, it is impossible to create.
